### ISSUE TITLE
[Research] Enable Custom Base-Types W/ Validators

### ISSUE THREAD

**Issue by dOrgJelli** [Web3-API/prototype#103]
Look more into what the implementation of custom base types w/ validators might looks like. An example of this might be a `CID` type from the IPFS package.

When importing / using these base types in other Web3APIs, custom validation code could be injected into the WASM package, fetched from the IPFS package itself.

**Comment by dOrgJelli** [comment]
Conversation brought over from Discord.

@Niraj-Kamdar: I really like the idea of allowing user to define custom type and marshalling. It will be really helpful when we get data in JSON format and would like to parse it in an appropriate shape. Currently we need to parse manually which isn't really a good experience for complex data."

@dOrgJelli: I definitely agree, but just not sure how to implement this quite yet. Would be very open to any proposals you may have.

Ideally any of our more "exotic" types that are built on-top of our base types (ex: BigInt, BigDecimal, JSON) could be specified using this framework.

One way of doing this that comes to mind is:
- custom-types must map to one or more base-types (ex: JSON -> String).
- for more complex custom types, you can specify structured objects using base-type properties.
- the generated code for this custom type will be a bit more complex. Its write/read logic will be broken up into steps. Let's look at read as an example:
- - step 1: the base-type is read (ex: base=reader.readString())
- - step 2: custom read logic is used to the new type (ex: json=CustomTypes.createJSON(base))
- - - this can be user-defined

Ideally we can have an "ecosystem" of these new base-types, and allow them to be easily embedded into people's modules to avoid having to send a "subinvoke" outside of the Wasm module (for speed reasons).

Projects like wasm-ld (build by LLVM (clang)) can be used to merge multiple Wasm modules together into a single binary output (using export linking).

Would love to continue this conversation here: https://github.com/polywrap/monorepo/issues/103

Reference links on wasm-ld:
https://lld.llvm.org/WebAssembly.html
https://aransentin.github.io/cwasm/

**Comment by Niraj-Kamdar** [comment]
Most of the APIs, we will use would be using JSON as response type. So, I think if we can add support for serializing and deserializing from/to JSON, it can be really helpful.

Ex: We make request to Coingecko API using HTTP plugin and get response in JSON and if we want to parse it into a custom type called PriceData then we need to parse JSON in assemblyscript using as-json library which isn't very good dev experience. 

I agree that we can add more methods for this custom type like `PriceData.fromJson(jsonString) -> PriceData` and `PriceData.toJson() -> jsonString`. In this case String would be the base type.

Although JSON is nice starting point, it shouldn't be just limited to JSON, in future user should be easily able to add support for other data type like xml, yaml, csv, etc.

**Comment by Niraj-Kamdar** [comment]
## Discussion on new Ethereum Address custom type
@Niraj-Kamdar: We should have Address type support
- It should also have function to convert lowercase address to canonical one
- it should validate the address during construction.
- custom comparators to compare string and Address with Address type. (using operator overriding)

Why is it needed? 
- Ideally I want to validate the entered address is valid before making the request to plugin
- experience to work with Address can be greatly improved by having this type

Although, we wouldn't be enforcing user to use Address type, they can still use plain string if they want to but it should be recommended to use Address for ethereum address.

@krisbitney: One issue with this is that we plan to support multiple blockchains, which may have different address protocols

@Niraj-Kamdar: 
- Yes I had that in mind, maybe we can have Ethereum_Address.

**Comment by Niraj-Kamdar** [comment]
## Discussion on custom-type in general

@Niraj-Kamdar:
- I think it would be useful if we can allow user to have custom types using extensions,  It should be as simple as installing a new npm/cargo dependency and passing it to client or something to add support for this new type 
- These would be 2 steps, we need to extend cli while building the wrapper and extend client while running the wrapper
- For wasm we need to inject this custom-type in schema-bind templates (template for templates haha)

@nerfZael: We can use Plugin custom-module as custom-types, Ex: if ethereum plugin has `Ethereum_Connection` and `Ethereum_Address` module, user can just create a custom-type using following syntax:
```ts
const a: Ethereum_Contract = Ethereum_Contract.create();
const b: Ethereum_Address = Ethereum_Address.create();
```
as opposed to what we have now:
```ts
const a: Ethereum_Contract = Ethereum_Query.createContract();
const b: Ethereum_Address = Ethereum_Query.createAddress(); 
```
@nerfZael and/or @Niraj-Kamdar:
- Benefit of having this is, we can easily implement this in plugin and anyone using this plugin can use it.
- This can sanitize input data while constructing type, support custom methods 
- It may be used in GraphQL schema 
- It can't be used as full-fledged type, it can't support operator overloading and won't be as performant as wasm native types

**Comment by Niraj-Kamdar** [comment]
I had an idea to implement this in a way that makes it extensible and scalable. We can provide a `Generic` scalar type. (We can even name it `Literal`) and have a Field Directive `annotate` with type property( Ex: `@annotate(type: String)`). 
```graphql
scalar Generic

directive @annotate(type: String) on FIELD   # in future we can annotate other things if needed

type CustomType {
  arg1: Generic! @annotate(type: "Map[String, Array[Int]]")
  arg2: Generic! @annotate(type: "BigDecimal")
  arg3: Generic! @annotate(type: "Address[ChainIdEnum]")
}
```


Developer can define any custom type here and need to provide a build-time parser that parses this type string into typeinfo types while we parse the schema, developer will also write bindings for the custom-type.

It would be nice if developer can create an npm package with parser and bindings to add support for any custom-type and user can specifies this module via cli or cli configs if they want to use it in their wrapper.

**Comment by Niraj-Kamdar** [comment]
I have a solution proposal for Data marshalling side of the custom-type:

## Prior Knowledge:
I think we can represent any user-defined custom type in one of the following category:
1) Scalar base type (int, float, string, bytes)
2) List collection  (Array/Set/Queue, etc)
3) Key-Value collection (Map/Object/Btree, etc)

## Naive Solution:
A simple approach would be to represent this as JSON/string type and serialize/deserialize it using JSON serializer.

## Performant Solution:
An alternate and more performant approach would be to use msgpack just like we are using for data-marshalling. 
Msgpack supports custom type extentions upto 128 different types which may be enough if we just need to cover all standard lib collection types but if we want to allow arbitrary amount of custom user-defined types then this can be very limiting so how can we still use msgpack but get away with the limitations of it.

Here's a way to solve this issue:

Before we start with solution, here's the example of how msgpack extention type look like: 
```
ext 8 stores an integer and a byte array whose length is upto (2^8)-1 bytes:
+--------+--------+--------+========+
|  0xc7  |XXXXXXXX|  type  |  data  |
+--------+--------+--------+========+
```

Here we can have common generic type with `0x00 as type` and we can have msgpack string (user defined type name) + payload msgpack (scalar/list collection (mapped to msgpack array)/ key-val collection (mapped to msgpack Map))
```
Our custom generic user-defined type
+--------+--------+--------+====================+=========+
|  0xc7  |XXXXXXXX|  type  |  custom type name  | payload |
+--------+--------+--------+====================+=========+

Note: bytes of custom type name + bytes of payload combined makes the whole bytes of data
```

String type name can further be improved by using hash/id of the type name and using msgpack fixstr, we can match this both client and wrapper side by keeping map of (hash/id -> type name) for user-defined types.

This won't limit us with the Msgpack's ext type limit of 128 types with the drawback of effectively decreasing available payload size but with efficient use of hash we can limit the waste of the payload space to certain extent.

This will be more efficient then the mapping types to string/JSON since msgpack is more performant and space efficient and. this won't add that much of development overhead either since we already have encoder/decoder for base categories (scalar, array, and map) which can be used to make any complex types.

**Comment by dOrgJelli** [comment]
I agree with this approach @Niraj-Kamdar, using a specific ext type (ex: 5) to denote custom user-defined polywrap type extension.

An example of what this could look like...

## MsgPack Base Types
string
uint
int
...

## Polywrap Base Type Extensions
generic map
generic set
json
bigint
...

## User Base Type Extensions
`#import { Address } into Eth from "ens/eth-address.polydevs.eth"`

where the resulting msgpack encoding would look like:
```
[0xc7]            # type extension
[XXXXXXXX]        # data length
[type]            # = 5; user-defined type extension wrapper
[uri_string]      # = "ens/eth-address...."; type extension wrapper URI
[typename_string] # = "Address"; type name
[data]            # length = XXXXXXXX
```
